gems/aws-sdk-core/lib/aws-sdk-core/instance_profile_credentials.rb

```
@@ -5,89 +5,68 @@
+  # be created with options provided.
   class InstanceProfileCredentials
     include CredentialProvider
     include RefreshingCredentials
 
-    # @api private
-    class Non200Response < RuntimeError; end
-
-    # @api private
-    class TokenRetrivalError < RuntimeError; end
-
-    # @api private
-    class TokenExpiredError < RuntimeError; end
-
-    # These are the errors we trap when attempting to talk to the
-    # instance metadata service.  Any of these imply the service
-    # is not present, no responding or some other non-recoverable
-    # error.
-    # @api private
-    NETWORK_ERRORS = [
-      Errno::EHOSTUNREACH,
-      Errno::ECONNREFUSED,
-      Errno::EHOSTDOWN,
-      Errno::ENETUNREACH,
-      SocketError,
-      Timeout::Error,
-      Non200Response
-    ].freeze
-
     # Path base for GET request for profile and credentials
     # @api private
-    METADATA_PATH_BASE = '/latest/meta-data/iam/security-credentials/'.freeze
-
-    # Path for PUT request for token
-    # @api private
-    METADATA_TOKEN_PATH = '/latest/api/token'.freeze
+    METADATA_PATH_BASE = '/latest/meta-data/iam/security-credentials/'
 
     # @param [Hash] options
-    # @option options [Integer] :retries (1) Number of times to retry
+    # @option options [Aws::EC2Metadata] :client A custom EC2 metadata client to
+    #   use for loading credentials from IMDSv2. If not provided, a default
+    #   {Aws::EC2Metadata} client will be constructed with passed options.
+    # @option options [String] :ec2_instance_profile_name (nil) When set, this
+    #   provider skips querying IMDS for the name of the active profile. When
+    #   `:instance_profile_name` are not configured directory, the following
+    #   locations will be searched before defaulting to `nil`:
+    #
+    #   * `ENV['AWS_EC2_INSTANCE_PROFILE_NAME']`
+    #   * `~/.aws/config`
+    #
+    # @option options [Integer] :retries (3) Number of times to retry
```

It's fine to default this to 3, however, ensure that in the chain, it still defaults to 1 - because we want the chain to fail fast (like in cases where EC2 is not used - we don't want to call a non-existent thing 3 times).

---

I think we should explicitly passed in `1` when we initialize this provider through the chain. It should be documented somewhere.

---

Hmm. I noticed the following documentation says...
 ```
    # @return [Integer] Number of times to retry when retrieving credentials
    #   from the instance metadata service. Defaults to 0 when resolving from
    #   the default credential chain ({Aws::CredentialProviderChain}).
```

(before my changes) - we had the retries to default as `1` in the init - but  when I look at chain ....it is set as `0`? I will dig into further documentation later tonight but let me kno.

https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/credential_provider_chain.rb#L36

---

It depends on how the retry logic is implemented, whether it checked greater than or greater than or equal to. We want to ensure in the chain that it is called once with no retry (1 attempt only). That would be 0 retries. Ensure the retry logic works that way if it doesn't already.

---

Gotcha, I'll update the docs to reflect what credentials chain have (aka `0`).